### PR TITLE
Implementation of InfluxDB Metrics (resolves #215)

### LINKED ISSUE #215: package metrics/influx
Although you could get by with StastD it would be nice to have native InfluxDB metrics reporting

### PR DESCRIPTION
**Pull request by DSchalla**
This is a work in progress PR for the implementation of the InfluxDB metrics package. The first commit contains the implementation for the counter writing to a InfluxDB backend with tag and field support.

**To Do:**
- [x] Counter Support
- [x] Counter Tests
- [x] Gauge Support
- [x] Gauge Tests
- [x] Histogram Support
- [x] Histogram Tests
- [x] Code Cleanup

Feel free to provide feedback regarding structure, tests and quality overall, its highly appreciated!

### CODE REVIEW

File: metrics/influxdb/influxdb.go [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+package influxdb
+
+import (
+	"time"
+
+	"github.com/go-kit/kit/metrics"
+	stdinflux "github.com/influxdata/influxdb/client/v2"
+)
+
+type influxdbCounter struct {
```
Comment (peterbourgon) [inline metrics/influxdb/influxdb.go]: You can just call this e.g. `counter`, influxdb is implicit from the package name.

File: metrics/influxdb/influxdb.go [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+	counter.bp.AddPoint(pt)
+}
+
+func (counter influxdbCounter) watch(client stdinflux.Client, bp stdinflux.BatchPoints, reportTicker <-chan time.Time) {
+	for {
+		select {
+		case <-reportTicker:
+			client.Write(bp)
+		}
+	}
```
Comment (peterbourgon) [inline metrics/influxdb/influxdb.go]:
```go
for range reportTicker {
    client.Write(bp)
}
```

File: metrics/influxdb/influxdb_test.go [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/go-kit/kit/metrics"
+	"github.com/go-kit/kit/metrics/influxdb"
+	stdinflux "github.com/influxdata/influxdb/client/v2"
+)
+
+func TestInfluxdbCounter(t *testing.T) {
```
Comment (peterbourgon) [inline metrics/influxdb/influxdb_test.go]: Likewise, `TestCounter` is sufficient.

File: metrics/influxdb/influxdb_test.go [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+		}
+		comparePoint(t, i, givenPoint, cl.Points[i])
+	}
+}
+
+func comparePoint(t *testing.T, i int, expected mockPoint, given stdinflux.Point) {
+	givenName := given.Name()
+	if givenName != expected.Name {
+		t.Errorf("Point %v invalid name, expected %v got %v", i, expected.Name, givenName)
+	}
```
Comment (peterbourgon) [inline metrics/influxdb/influxdb_test.go]:
```go
if want, have := expected.Name, given.Name(); want != have {
    t.Errorf("want %q, have %q", want, have)
}
```

File: metrics/influxdb/influxdb_test.go [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+
+	tagsEqual := reflect.DeepEqual(expected.Tags, given.Tags())
+	if !tagsEqual {
+		t.Errorf("Point %v invalid tags, expected %v got %v", i, expected.Tags, given.Tags())
+	}
```
Comment (peterbourgon) [inline metrics/influxdb/influxdb_test.go]:
```go
if want, have := expected.Tags, given.Tags(); !reflect.DeepEqual(want, have) {
    t.Errorf("want %v, have %v", want, have)
}
```

File: metrics/influxdb/influxdb_test.go [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+
+	fieldsEqual := reflect.DeepEqual(expected.Fields, given.Fields())
+	if !fieldsEqual {
+		t.Errorf("Point %v invalid fields, expected %v got %v", i, expected.Fields, given.Fields())
+	}
```
Comment (peterbourgon) [inline metrics/influxdb/influxdb_test.go]:
```go
if want, have := expected.Fields, given.Fields(); !reflect.DeepEqual(want, have) {
     t.Errorf("point %d: want %v, have %v", i, want, have)
}
```

File: metrics/influxdb/influxdb.go [hunk 7]
```diff
@@ -0,0 +1,78 @@
+// Package influxdb implements a Influxdb backend for package metrics.
```
Comment (peterbourgon) [inline metrics/influxdb/influxdb.go]: Minor nit: `s/Influxdb/InfluxDB/g`

**Comment by DSchalla** [comment]
Thanks for the feedback @peterbourgon! The last commit introduced the proposed changes, I asked myself if the prefix struct naming should be removed in other packages aswell? I checked Prometheus and Statsd, they both used the same convention, while dogstatsd does not.

**Comment by peterbourgon** [comment]
Yes, that's a problem in those packages! I filed https://github.com/go-kit/kit/issues/266 :)

File: metrics/influxdb/influxdb.go [hunk 8]
```diff
@@ -0,0 +1,254 @@
...
+import (
+	"fmt"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/codahale/hdrhistogram"
+	"github.com/go-kit/kit/metrics"
+	stdinflux "github.com/influxdata/influxdb/client/v2"
+)
```
Comment (peterbourgon) [inline metrics/influxdb/influxdb.go]:
```go
import (
    // stdlib

    // third-party

    // go-kit/kit/*
)
```

File: metrics/influxdb/influxdb.go [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+type counter struct {
+	key    string
+	tags   []metrics.Field
+	fields []metrics.Field
+	value  uint64
+	bp     stdinflux.BatchPoints
+}
+
+// NewCounter returns a Counter that writes values in the reportInterval
+// to the given InfluxDB client, utilizing batching
```
Comment (peterbourgon) [inline metrics/influxdb/influxdb.go]: `batching.`

File: metrics/influxdb/influxdb.go [hunk 10]
```diff
@@ -0,0 +1,254 @@
...
+func NewCounter(client stdinflux.Client, bp stdinflux.BatchPoints, key string, tags []metrics.Field, reportInterval time.Duration) metrics.Counter {
+	return NewCounterTick(client, bp, key, tags, time.Tick(reportInterval))
+}
+
+// NewCounterTick is the same as NewCounter, but allows the user to pass a own
+// channel to trigger the write process to the client
```
Comment (peterbourgon) [inline metrics/influxdb/influxdb.go]: `client.`

File: metrics/influxdb/influxdb.go [hunk 11]
```diff
@@ -0,0 +1,254 @@
...
+		key:   key,
+		tags:  tags,
+		value: 0,
+		bp:    bp,
+	}
+	go watch(client, bp, reportTicker)
+	return c
+}
+
+func (c *counter) Name() string {
```
Comment (peterbourgon) [inline metrics/influxdb/influxdb.go]:
```go
// Name implements the whatever interface.
```

File: metrics/influxdb/influxdb.go [hunk 12]
```diff
@@ -0,0 +1,254 @@
...
+	return c.key
+}
+
+func (c *counter) With(field metrics.Field) metrics.Counter {
```
Comment (peterbourgon) [inline metrics/influxdb/influxdb.go]: Doc comment

File: metrics/influxdb/influxdb.go [hunk 13]
```diff
@@ -0,0 +1,254 @@
...
+	go h.rotateLoop(1 * time.Minute)
+	return h
+}
+
+func (h *histogram) Name() string {
+	return h.key
+}
+
+func (h *histogram) With(field metrics.Field) metrics.Histogram {
+
```
Comment (peterbourgon) [inline metrics/influxdb/influxdb.go]: Extraneous newline

File: metrics/influxdb/influxdb.go [hunk 14]
```diff
@@ -0,0 +1,254 @@
...
+	for _, field := range g.fields {
+		fields[field.Key] = field.Value
+	}
+	fields["value"] = g.value
+	pt, _ := stdinflux.NewPoint(g.key, tags, fields, time.Now())
+	g.bp.AddPoint(pt)
+}
+
+// The implementation from histogram is taken from metrics/expvar
```
Comment (peterbourgon) [inline metrics/influxdb/influxdb.go]: Does it make sense to bring this out to a new package `metrics/internal/histogram`?

File: metrics/influxdb/influxdb_test.go [hunk 15]
```diff
@@ -0,0 +1,347 @@
...
+import (
+	"reflect"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/go-kit/kit/metrics"
+	"github.com/go-kit/kit/metrics/influxdb"
+	stdinflux "github.com/influxdata/influxdb/client/v2"
+)
```
Comment (peterbourgon) [inline metrics/influxdb/influxdb_test.go]:
```go
import (
    // stdlib

    // third-party

    // go-kit/kit/*
)
```

**Comment by peterbourgon** [comment]
Back to you, @DSchalla! Just a couple of small-ish things.

**Comment by DSchalla** [comment]
Sure, I will update the PR tomorrow. Regarding the histogram, I dont know. It might make sense, since at least 2 providers are using it already, but then again one has to come up with a good solution which still allows fields for one provider, but doesnt enforce it for other providers. I am open for suggestions tho.

**Comment by peterbourgon** [comment]
OK, we can tackle extracting the histogram in perhaps #263

**Comment by DSchalla** [comment]
If there is anything left to do in order to merge the PR, please just give me a hint.

**Comment by peterbourgon** [comment]
OK. I am traveling to and speaking at GlueCon and will be unavailable for the next ~2 days. I'll pick this up again after that, if it's OK with you.

**Comment by DSchalla** [comment]
Sure, thats fine for me. Have fun!

**Comment by peterbourgon** [comment]
Brilliant! Thanks so much!

**Pull request merged**